### PR TITLE
[MLIR][Presburger] Add Gram-Schmidt

### PR DESCRIPTION
**Pull request by Abhinav271828**
Implement Gram-Schmidt orthogonalisation for the FracMatrix class.
This requires dotProduct, which has been added as a util.

### CODE REVIEW

**Review by superty — COMMENTED** [review]
Please use slightly longer lines for comments, you can make use of the full 80 lines of space that are available. You can just write long lines and let the formatter wrap them. You don't have to worry about keeping them short.

File: mlir/lib/Analysis/Presburger/Matrix.cpp [hunk 1]
```diff
@@ -548,4 +548,27 @@ Fraction FracMatrix::determinant(FracMatrix *inverse) const {
...
+    // For each vector (row) in the matrix,
+    // subtract its unit projection along
+    // each of the previous vectors.
+    // This ensures that it has no component
+    // in the direction of any of the
+    // previous vectors.
+    for (unsigned i = 1; i < getNumRows(); i++) {
+        for (unsigned j = 0; j < i; j++) {
+            projectionScale = dotProduct(orth.getRow(i), orth.getRow(j)) /
+                              dotProduct(orth.getRow(j), orth.getRow(j));
```
Comment (Superty) [inline mlir/lib/Analysis/Presburger/Matrix.cpp:569]: what happens if some row becomes zero?

File: mlir/unittests/Analysis/Presburger/MatrixTest.cpp [hunk 2]
```diff
@@ -310,3 +310,18 @@ TEST(MatrixTest, intInverse) {
...
+TEST(MatrixTest, gramSchmidt) {
+  FracMatrix mat = makeFracMatrix(3, 5, {{Fraction(3, 1), Fraction(4, 1), Fraction(5, 1), Fraction(12, 1), Fraction(19, 1)},
+                                         {Fraction(4, 1), Fraction(5, 1), Fraction(6, 1), Fraction(13, 1), Fraction(20, 1)},
+                                         {Fraction(7, 1), Fraction(8, 1), Fraction(9, 1), Fraction(16, 1), Fraction(24, 1)}});
+
+  FracMatrix gramSchmidt = makeFracMatrix(3, 5,
+         {{Fraction(3, 1),     Fraction(4, 1),     Fraction(5, 1),    Fraction(12, 1),     Fraction(19, 1)},
+          {Fraction(142, 185), Fraction(383, 555), Fraction(68, 111), Fraction(13, 185),   Fraction(-262, 555)},
+          {Fraction(53, 463),  Fraction(27, 463),  Fraction(1, 463),  Fraction(-181, 463), Fraction(100, 463)}});
+
```
Comment (Superty) [inline mlir/unittests/Analysis/Presburger/MatrixTest.cpp:323]: add a test for the zero case please

File: mlir/unittests/Analysis/Presburger/MatrixTest.cpp [hunk 3]
```diff
@@ -310,3 +310,18 @@ TEST(MatrixTest, intInverse) {
...
+  FracMatrix gs = mat.gramSchmidt();
+
+  EXPECT_EQ_FRAC_MATRIX(gs, gramSchmidt);
```
Comment (Superty) [inline mlir/unittests/Analysis/Presburger/MatrixTest.cpp:326]: write a check that all the returned vectors are actually orthogonal.

**Review by superty — CHANGES_REQUESTED** [review]

File: mlir/lib/Analysis/Presburger/Matrix.cpp [hunk 4]
```diff
@@ -548,4 +548,29 @@ Fraction FracMatrix::determinant(FracMatrix *inverse) const {
     determinant *= m.at(i, i);
 
   return determinant;
+}
+
+FracMatrix FracMatrix::gramSchmidt() const {
+  bool linIndep =
+      (nRows < nColumns) || (nRows == nColumns && determinant(nullptr) != 0);
```
Comment (Superty) [inline mlir/lib/Analysis/Presburger/Matrix.cpp:555]: you can skip the argument to determinant, it should default to nullptr right?

File: mlir/lib/Analysis/Presburger/Matrix.cpp [hunk 5]
```diff
@@ -548,4 +548,29 @@ Fraction FracMatrix::determinant(FracMatrix *inverse) const {
...
+  // the orthogonalised version.
+  FracMatrix orth(*this);
+
+  // For each vector (row) in the matrix,
+  // subtract its unit projection along
+  // each of the previous vectors.
+  // This ensures that it has no component
+  // in the direction of any of the
+  // previous vectors.
+  for (unsigned i = 1; i < getNumRows(); i++) {
```
Comment (Superty) [inline mlir/lib/Analysis/Presburger/Matrix.cpp:568]: for (i = 1, e = getNumRows(); i < e; ++i)

Comment (Superty) [reply mlir/lib/Analysis/Presburger/Matrix.cpp:568]: https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop

Comment (Superty) [reply mlir/lib/Analysis/Presburger/Matrix.cpp:568]: Please apply to all loops.

File: mlir/unittests/Analysis/Presburger/MatrixTest.cpp [hunk 6]
```diff
@@ -310,3 +310,30 @@ TEST(MatrixTest, intInverse) {
...
+       {Fraction(53, 463), Fraction(27, 463), Fraction(1, 463),
+        Fraction(-181, 463), Fraction(100, 463)}});
+
+  FracMatrix gs = mat.gramSchmidt();
+
+  EXPECT_EQ_FRAC_MATRIX(gs, gramSchmidt);
+  for (unsigned i = 0; i < 3u; i++)
+    for (unsigned j = i + 1; j < 3u; j++)
+      EXPECT_EQ(dotProduct(gramSchmidt.getRow(i), gramSchmidt.getRow(j)), 0);
+}
```
Comment (Superty) [inline mlir/unittests/Analysis/Presburger/MatrixTest.cpp:339]: Please add a few more tests.

Comment (Superty) [reply mlir/unittests/Analysis/Presburger/MatrixTest.cpp:339]: Maybe one can be the identity matrix

**Review by superty — CHANGES_REQUESTED** [review]

File: mlir/unittests/Analysis/Presburger/MatrixTest.cpp (hunk 6 above)
Comment (Abhinav271828) [reply mlir/unittests/Analysis/Presburger/MatrixTest.cpp:377]: added some more, is this enough?

File: mlir/lib/Analysis/Presburger/Matrix.cpp [hunk 7]
```diff
@@ -548,4 +548,27 @@ Fraction FracMatrix::determinant(FracMatrix *inverse) const {
     determinant *= m.at(i, i);
 
   return determinant;
+}
+
+FracMatrix FracMatrix::gramSchmidt() const {
+  bool linIndep =
+      (nRows < nColumns) || (nRows == nColumns && determinant() != 0);
+  assert(linIndep && "the vectors must be linearly independent!");
```
Comment (Superty) [inline mlir/lib/Analysis/Presburger/Matrix.cpp:556]: less vectors than the dimension does not imply linear independence.

File: mlir/unittests/Analysis/Presburger/MatrixTest.cpp (hunk 6 above)
Comment (Superty) [reply mlir/unittests/Analysis/Presburger/MatrixTest.cpp:377]: For the one where you don't compare against a reference martix, checking if the result has the same span is too much effort, but at least check if the vectors stayed linearly independent (determinant != 0). BTW, if you don't need this function for less that nDim vectors you can just add an assert for that.

**Comment by Superty** [comment]
I prefer an assert statement that clearly says what actually happened

"Some row became zero! Inputs to this function must be linearly independent."

File: mlir/lib/Analysis/Presburger/Matrix.cpp [hunk 8]
```diff
@@ -548,4 +548,25 @@ Fraction FracMatrix::determinant(FracMatrix *inverse) const {
...
+  // the orthogonalised version.
+  FracMatrix orth(*this);
+
+  // For each vector (row) in the matrix, subtract its unit
+  // projection along each of the previous vectors.
+  // This ensures that it has no component in the direction
+  // of any of the previous vectors.
+  for (unsigned i = 1, e = getNumRows(); i < e; i++) {
+    for (unsigned j = 0; j < i; j++) {
+      Fraction jDotProd = dotProduct(orth.getRow(j), orth.getRow(j));
```
Comment (Superty) [inline mlir/lib/Analysis/Presburger/Matrix.cpp:564]: "jDotProd" -> `jNormSquared`

**Pull request merged**